pkg/registry/bundlegraphloader_test.go

```diff
@@ -0,0 +1,345 @@
+							BundleKey{CsvName: "etcdoperator.v0.9.3", Version: "0.9.3"}: nil,
+						}},
+				},
+			},
+			newDefaultChannel: "alpha",
+		},
+		{ // Add a bundle to an empty graph
+			fail:  false,
+			graph: Package{},
+			bundle: Bundle{
+				Name:    "etcdoperator.v0.9.3",
+				Package: "etcd",
+				csv: &ClusterServiceVersion{
+					Spec: json.RawMessage(`
+						{ 
+						"version": "0.9.3"
+						}`),
+				},
+				Channels: []string{"alpha"},
+			},
+			expectedGraph: &Package{
+				Name:           "etcd",
+				DefaultChannel: "alpha",
+				Channels: map[string]Channel{
+					"alpha": {Head: BundleKey{CsvName: "etcdoperator.v0.9.3", Version: "0.9.3"},
+						Nodes: map[BundleKey]map[BundleKey]struct{}{
+							BundleKey{CsvName: "etcdoperator.v0.9.3", Version: "0.9.3"}: nil,
+						}},
+				},
+			},
+			newDefaultChannel: "alpha",
+		},
+		{ // Add a bundle in skippatch mode
+			fail: false,
+			graph: Package{
+				Name:           "etcd",
+				DefaultChannel: "alpha",
+				Channels: map[string]Channel{
+					"alpha": {Head: BundleKey{CsvName: "etcdoperator.v0.9.2", Version: "0.9.2"},
+						Nodes: map[BundleKey]map[BundleKey]struct{}{
+							BundleKey{CsvName: "etcdoperator.v0.6.1", Version: "0.6.1"}: {},
+							BundleKey{CsvName: "etcdoperator.v0.9.0", Version: "0.9.0"}: {BundleKey{CsvName: "etcdoperator.v0.6.1", Version: "0.6.1"}: {}},
+							BundleKey{CsvName: "etcdoperator.v0.9.2", Version: "0.9.2"}: {BundleKey{CsvName: "etcdoperator.v0.9.0", Version: "0.9.0"}: {},
+								BundleKey{CsvName: "etcdoperator.v0.9.1"}: {}},
+						}},
+
+					"beta": {Head: BundleKey{CsvName: "etcdoperator.v0.9.0", Version: "0.9.0"},
+						Nodes: map[BundleKey]map[BundleKey]struct{}{
+							BundleKey{CsvName: "etcdoperator.v0.6.1", Version: "0.6.1"}: {},
+							BundleKey{CsvName: "etcdoperator.v0.9.0", Version: "0.9.0"}: {BundleKey{CsvName: "etcdoperator.v0.6.1", Version: "0.6.1"}: {}},
+						}},
+
+					"stable": {Head: BundleKey{CsvName: "etcdoperator.v0.9.2", Version: "0.9.2"},
+						Nodes: map[BundleKey]map[BundleKey]struct{}{
+							BundleKey{CsvName: "etcdoperator.v0.6.1", Version: "0.6.1"}: {},
+							BundleKey{CsvName: "etcdoperator.v0.9.0", Version: "0.9.0"}: {BundleKey{CsvName: "etcdoperator.v0.6.1", Version: "0.6.1"}: {}},
+							BundleKey{CsvName: "etcdoperator.v0.9.2", Version: "0.9.2"}: {BundleKey{CsvName: "etcdoperator.v0.9.0", Version: "0.9.0"}: {},
+								BundleKey{CsvName: "etcdoperator.v0.9.1"}: {}},
+						}},
+				},
+			},
+			bundle: Bundle{
+				Name:    "etcdoperator.v0.9.3",
+				Package: "etcd",
+				csv: &ClusterServiceVersion{
+					Spec: json.RawMessage(`
+							{
+							"version": "0.9.3"
+							}`),
+				},
+				Channels: []string{"alpha", "stable"},
+			},
+			skipPatch: true,
+			expectedGraph: &Package{
+				Name:           "etcd",
+				DefaultChannel: "alpha",
+				Channels: map[string]Channel{
+					"alpha": {Head: BundleKey{CsvName: "etcdoperator.v0.9.3", Version: "0.9.3"},
```

Comment:
what's the right graph for skippatch mode? there shouldn't be 0.9.0 or 0.9.2 entry, right? 

Should alpha just be:

0.9.3 -> (0.9.0, 0.9.1, 0.9.2, 0.6.1)

0.9.3 skips 0.9.2, so it should skip 0.9.1 (skipped by 0.9.2) transitively
0.9.3 skips 0.9.0, so it should skip 0.6.1 (replaced by 0.9.0) transitively (I think this is arguable)

either way I don't think there should be an entry for 0.9.2, since it's definitively skipped in skippatch mode.

Does this definition of skippatch mode make sense: "If I add 0.Y.Z in skip patch mode, anything that would have updated to any 0.Y.W should instead update directly to 0.Y.Z"?

Reply:
Spoke a little about this with Evan, my opinion is that non-latest z versions should not be considered "real" nodes in the graph in skip-patch mode.

Updated the pr to reflect that in the bundle updater.